### PR TITLE
Assay file improvements

### PR DESCRIPTION
**Pull request by bedroesb**
### Examples:

a_RIGW1_BLOCK>PLOT.txt:

Sample Name | Protocol REF | Assay Name | Derived Data File
-- | -- | -- | --
26212 | phenotyping | 2990931-2008 | d_RIGW1_BLOCK>PLOT.txt
26212 | phenotyping | 2990931-2007 | d_RIGW1_BLOCK>PLOT.txt
26212 | phenotyping | 2990931-2009 | d_RIGW1_BLOCK>PLOT.txt
26212 | phenotyping | 2990931-2006 | d_RIGW1_BLOCK>PLOT.txt

a_POPYOMICS-POP2-F_VIRTUAL_TRIAL.txt:

Sample Name | Protocol REF | Assay Name | Derived Data File
-- | -- | -- | --
101-74 | phenotyping | 2889403-2004 | d_POPYOMICS-POP2-F_VIRTUAL_TRIAL.txt
101-74 | phenotyping | 2889403-2005 | d_POPYOMICS-POP2-F_VIRTUAL_TRIAL.txt
661300224 | phenotyping | 2889073-2005 | d_POPYOMICS-POP2-F_VIRTUAL_TRIAL.txt
661300224 | phenotyping | 2889073-2004 | d_POPYOMICS-POP2-F_VIRTUAL_TRIAL.txt

a_VIB_study___40_plant.txt:

Sample Name | Protocol REF | Assay Name | Derived Data File
-- | -- | -- | --
pot_1 | phenotyping | P_6583 | d_VIB_study___40_plant.txt
pot_10 | phenotyping | P_6592 | d_VIB_study___40_plant.txt
pot_100 | phenotyping | P_6463 | d_VIB_study___40_plant.txt
pot_101 | phenotyping | P_6464 | d_VIB_study___40_plant.txt

a_BTH_Clermont-Ferrand_2000_SetB1_REPLICATION.txt:

Sample Name | Protocol REF | Assay Name | Derived Data File
-- | -- | -- | --
1 | phenotyping | 2763979-2000 | d_BTH_Clermont-Ferrand_2000_SetB1_REPLICATION.txt
2 | phenotyping | 2763968-2000 | d_BTH_Clermont-Ferrand_2000_SetB1_REPLICATION.txt
CF00188itkt | phenotyping | 2763922-2000 | d_BTH_Clermont-Ferrand_2000_SetB1_REPLICATION.txt
CF00188itkt | phenotyping | 2763977-2000 | d_BTH_Clermont-Ferrand_2000_SetB1_REPLICATION.txt

### CODE REVIEW

**Comment by cpommier** [comment]
@bedroesb I am generating my datafiles with that

**Comment by bedroesb** [comment]
I am still trying to figure out the material - protocol - source - sample - assay object hierarchy in ISA to generate the assay files how it should be

**Comment by proccaserra** [comment]
@bedroesb what is wrong with the current one as available in the original code (I mean besides the new request that came up to break down by observational units as we went along) ?

File: brapi_to_isa.py [hunk 1]
```diff
@@ -150,39 +150,31 @@ def create_study_sample_and_assay(client, brapi_study_id, isa_study,  sample_col
...
+            # !!!: fix isatab.py to access other protocol_type values to enable Assay Tab serialization
+        phenotyping_process = Process(executes_protocol=phenotyping_protocol)
         
+        material = Material(name="extract-{}".format(obs_unit["observationUnitDbId"]))
+        material.type = "test"
+        c = Characteristic(category=OntologyAnnotation(term="Observation Unit Type"),
+                                value=OntologyAnnotation(term=obslvl,
+                                                                    term_source="",
+                                                                    term_accession=""))
+        material.characteristics.append(c)
```
Comment (bedroesb) [inline brapi_to_isa.py]: @proccaserra I still didn't manage to get the material (+characteristics part) working to have a characteristics column in the assay file...

Comment (bedroesb) [reply brapi_to_isa.py]: If you would have some time to look at my code, that would be nice. Since I am not confident about the protocol/sample part

Comment (proccaserra) [reply brapi_to_isa.py]: @bedroesb I see what is happening. You created a material which is on his own, not linked to either a Source or a Sample. The key here is to be sure about the object you'll add the Characteristics to. Also if you have only protocol even creating  'Sample' , the characteristics will be added to that one sample. Upon serialisation to ISA-Tab, the Characteristics will appear in the s_study file.
So the thing to clarify is whether you need to quality a Sample with a new Characteristic or a New ISA Material (e.g Extract) with that Characteristics.

Comment (bedroesb) [reply brapi_to_isa.py]: @proccaserra Ok thank you, I think that with this I can figure it out.
Sorry I didn't yet respond to you mail, but I was like, how can I not figure this out myself ? ;)

**Comment by bedroesb** [comment]
@cpommier what do you mean ?

File: brapi_to_isa.py (hunk 1 above)
Comment (proccaserra) [reply brapi_to_isa.py]: @bedroesb  not pb at all. thx. I am going over the code now.

Comment (PapoutsoglouE) [reply brapi_to_isa.py]: @proccaserra Is it not possible to serialize characteristics referring to a Sample (not an Extract!) as columns of an assay file? I'm not sure I understand.

Comment (bedroesb) [reply brapi_to_isa.py]: @proccaserra the endpoints that I use to test the script with:
- `python brapi_to_isa.py -t all -e https://pippa.psb.ugent.be/BrAPIPPA/brapi/v1/`
- `python brapi_to_isa.py -s RIGW1 -e https://urgi.versailles.inra.fr/gnpis-core-srv/brapi/v1/`
- `python brapi_to_isa.py -t 24 -e https://urgi.versailles.inra.fr/gnpis-core-srv/brapi/v1/`

and a really big one that contains treatments and I terminate in the process: 
- `python brapi_to_isa.py -t 6 -e https://urgi.versailles.inra.fr/gnpis-core-srv/brapi/v1/`

**Comment by bedroesb** [comment]
Missing columns at the moment based on isa validator:
- Characteristics[Observation Unit Type]
- Parameter Value[Collection Date]
- Parameter Value[Sample Description]
- Raw Data File

**Comment by bedroesb** [comment]
Issues that still need to be solved based on isa validator:
```json
"message": "Missing Protocol declaration",
            "supplemental": "Protocol(s) of type ['Phenotyping', 'Sampling'] defined in the ISA-configuration expected as a between 'Sample Name' and 'Assay Name' but has not been found, in the file 'a_VIB_study___55_plant.txt'",
            "code": 1007
```
```json
"message": "Missing Protocol declaration",
            "supplemental": "Protocol(s) of type ['Data Transformation'] defined in the ISA-configuration expected as a between 'Assay Name' and 'Derived Data File' but has not been found, in the file 'a_VIB_study___55_plant.txt'",
            "code": 1007
```

**Comment by bedroesb** [comment]
Will open new pull request/issue with final to do list with things that need to be done, and were I will need assistance with.

**Pull request merged**